### ISSUE TITLE
Debug flaky `TestStreamForServer` test

### ISSUE THREAD

**Issue by emcfarlane** [connectrpc/connect-go#626]
**Describe the bug**

On [bump golang.org/x/net from 0.16.0 to 0.17.0](https://github.com/connectrpc/connect-go/actions/runs/6751140307/job/18354726039) tests can fail on `Send` reporting a closed request body error:
```
--- FAIL: TestStreamForServer (0.25s)
    --- FAIL: TestStreamForServer/get-spec (0.37s)
        connect_ext_test.go:1562: 
            assertion:	assert.Nil
            got:	unknown: EOF
```

**To Reproduce**

```bash
go test -run=TestStreamForServer -count=1000
```

**Additional context**

Tracing down the calls:
- [`Write`](https://github.com/connectrpc/connect-go/blob/2be1555cbf36879b6b1197cd09bc6e4cbe10290f/duplex_http_call.go#L107) returns an `io.EOF` error from a close on the request body (`io.ErrClosedPipe`)
- [`Close`](https://github.com/golang/net/blob/b225e7ca6dde1ef5a5ae5ce922861bda011cfabd/http2/transport.go#L449) is called by the transport
- [`abortStream`](https://github.com/golang/net/blob/b225e7ca6dde1ef5a5ae5ce922861bda011cfabd/http2/transport.go#L2934) from a process reset stream frame with the error:
```
stream error: stream ID 3; NO_ERROR; received from peer:
```

Unrelated to changes in x/net 0.17.0. Occurs in the latest v1.12 release.

**Comment by jhump** [comment]
Hmm. Do you know _why_ the stream is being aborted? That error message ("NO_ERROR received from peer") suggests the server sent a GOAWAY frame with error code NO_ERROR, which is supposed to mean that the server is gracefully shutting down. Is there a possible race with an RPC operation completing the server being shutdown prematurely? (Like maybe there's a `defer` function that should instead be a `t.Cleanup` function somewhere?)

**Comment by jhump** [comment]
> The PR goes into the details.

Okay, moving this convo to the PR.

**Comment by emcfarlane** [comment]
Would be good to discuss the intended error behaviour here. I think that `Send(nil)` should always succeed when initializing the request headers. From the `get-spec` test above, the behaviour that `Send(nil)` could return `io.EOF` is surprising to me and I'd expect shouldn't happen before calling `CloseRequest`:
https://github.com/connectrpc/connect-go/blob/20b5723aa3c09d35c0ce1b3b120796b5382388fd/connect_ext_test.go#L1561-L1563

Currently `Send(nil)` wait's for request body read. The behaviour of `io.Reader` states:
> If len(p) == 0, Read should always return n == 0. It may return a non-nil error if some error condition is known, such as EOF.
>
> Implementations of Read are discouraged from returning a zero byte count with a nil error, except when len(p) == 0. Callers should treat a return of 0 and nil as indicating that nothing happened; in particular it does not indicate EOF.

We could also remove the call to `Write` and instead call `ensureRequestMade()` to avoid capturing `Close` errors. I think this behaviour would follow the godoc for `Send(nil)`.
https://github.com/connectrpc/connect-go/blob/20b5723aa3c09d35c0ce1b3b120796b5382388fd/duplex_http_call.go#L98-L101

**Comment by jhump** [comment]
> I think this behaviour would follow the godoc for `Send(nil)`.

For reference, the actual Go docs for `Send` are below.

_connect.ClientStreamForClient_:
> Send a message to the server. The first call to Send also sends the request headers.
>
> If the server returns an error, Send returns an error that wraps [io.EOF](https://pkg.go.dev/io#EOF). Clients should check for case using the standard library's [errors.Is](https://pkg.go.dev/errors#Is) and unmarshal the error using CloseAndReceive.

_connect.BidiStreamForClient_:
> Send a message to the server. The first call to Send also sends the request headers. To send just the request headers, without a body, call Send with a nil pointer.
>
> If the server returns an error, Send returns an error that wraps [io.EOF](https://pkg.go.dev/io#EOF). Clients should check for EOF using the standard library's [errors.Is](https://pkg.go.dev/errors#Is) and call Receive to retrieve the error.

I see nothing wrong with the described behavior (aside from the fact that this blurb about a nil argument is omitted from the client-stream docs). They do not suggest anything special about the use of nil other than sending request headers without a body message.  They clearly state that the method returns `io.EOF` if the server returns an error, regardless of whether the argument is nil or not.

Special-casing the behavior (the opposite of uniformity) can make an API harder to reason about and harder to use (and potentially harder to implement). I don't see the value of doing so 🤷 

> Would be good to discuss the intended error behaviour here.

This issue is about fixing a flaky test. And the flaky test could be remedied _without_ changing any behavior. If you want to discuss changing the behavior (and changing/refining the specs), I think that calls for creating a separate issue, instead of doing it here. Since the test can be de-flaked without changing the behavior (by instead relaxing assertions), that means the current PR doesn't need to be blocked on the outcome of such a discussion.

**Comment by jhump** [comment]
Your last comment over in the PR helps me understand the situation more clearly. I think I misunderstood your suggestion to never return errors for `Send(nil)`, and I definitely was not grok'ing this:

> There are client `Send`s that are never matched by a server `Receive`.

The above doesn't seem aberrant to me since a send doesn't _have_ to be received. But I misunderstood how this caused the issue. I think I would have worded it: "the handler closes the stream with a non-error response without accepting any messages".

I think I understand the issue better now. Apologies for being dense about it before. I was interpreting "behavior change" and your previous proposal too broadly. But I do see now that `Send(nil)` should indeed return nil, even if the server has closed the stream, under the following conditions:
1. If zero actual messages have been sent or received (so this send is clearly intended to only send headers).
2. If the error returned from `Receive` would be `io.EOF`. In other words: the stream is closed with **no** error.

**Comment by emcfarlane** [comment]
I've changed the fix to check if the `Send(nil)` caused the request to be sent. Only then do we avoid writing the nil checking the state of the io.Pipe. Afterwards any `Send(nil)` will write. 

Sorry for the confusion. I've revised the description on the ticket to hopefully clarify the issue.

**Comment by jhump** [comment]
> Sorry for the confusion.

Nah, not your fault. I think I was splitting my attention too many ways and not spending enough time to truly comprehend the issue here before providing that earlier feedback.
